### ISSUE TITLE
Cluster/Remoting: ACKs misaligned (?)

### ISSUE THREAD

**Issue by ktoso** [akka/akka#19780]
Core issue:

```
February 12th 2016, 04:00:59.000	akka.remote.Remoting	ip-10-0-200-7	ERROR	cluster	Association to [akka.tcp://application@xx.xx.xx.xx:2551] with UID [656067421] irrecoverably failed. Quarantining address. java.lang.IllegalStateException: Error encountered while processing system message acknowledgement buffer: [-1 {}] ack: ACK[7, {}]t
	at akka.remote.ReliableDeliverySupervisor$$anonfun$receive$1.applyOrElse(Endpoint.scala:288) ~[com.typesafe.akka.akka-remote_2.11-2.4.1.jar:na]t
	at akka.actor.Actor$class.aroundReceive(Actor.scala:480) ~[com.typesafe.akka.akka-actor_2.11-2.4.1.jar:na]t	at akka.remote.ReliableDeliverySupervisor.aroundReceive(Endpoint.scala:189) ~[com.typesafe.akka.akka-remote_2.11-2.4.1.jar:na]t
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526) [com.typesafe.akka.akka-actor_2.11-2.4.1.jar:na]t
	at akka.actor.ActorCell.invoke(ActorCell.scala:495) [com.typesafe.akka.akka-actor_2.11-2.4.1.jar:na]t
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257) [com.typesafe.akka.akka-actor_2.11-2.4.1.jar:na]t
	at akka.dispatch.Mailbox.run(Mailbox.scala:224) [com.typesafe.akka.akka-actor_2.11-2.4.1.jar:na]t
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234) [com.typesafe.akka.akka-actor_2.11-2.4.1.jar:na]t
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [org.scala-lang.scala-library-2.11.7.jar:na]t
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [org.scala-lang.scala-library-2.11.7.jar:na]t
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [org.scala-lang.scala-library-2.11.7.jar:na]t
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [org.scala-lang.scala-library-2.11.7.jar:na]t
Caused by: java.lang.IllegalArgumentException: Highest SEQ so far was -1 but cumulative ACK is 7t
	at akka.remote.AckedSendBuffer.acknowledge(AckedDelivery.scala:103) ~[com.typesafe.akka.akka-remote_2.11-2.4.1.jar:na]t
	at akka.remote.ReliableDeliverySupervisor$$anonfun$receive$1.applyOrElse(Endpoint.scala:284) ~[com.typesafe.akka.akka-remote_2.11-2.4.1.jar:na]t
	... 11 common frames omittedt
```

**Comment by ktoso** [comment]
(Moved it to 2.4.3 to keep a close eye on it)

**Comment by marekzebrowski** [comment]
I can confirm that behavior on EC2. I started with a cluster of 3 nodes, then quickly started new 8 nodes that try to connect.
in effect:
```
 at akka.actor.ActorCell.invoke(ActorCell.scala:495) [akka-actor_2.11-2.4.2.jar:na]
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257) [akka-actor_2.11-2.4.2.jar:na]
        at akka.dispatch.Mailbox.run(Mailbox.scala:224) [akka-actor_2.11-2.4.2.jar:na]
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234) [akka-actor_2.11-2.4.2.jar:na]
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [scala-library-2.11.7.jar:na]
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [scala-library-2.11.7.jar:na]
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [scala-library-2.11.7.jar:na]
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [scala-library-2.11.7.jar:na]
Caused by: java.lang.IllegalArgumentException: Highest SEQ so far was -1 but cumulative ACK is 5
        at akka.remote.AckedSendBuffer.acknowledge(AckedDelivery.scala:103) ~[akka-remote_2.11-2.4.2.jar:na]
        at akka.remote.ReliableDeliverySupervisor$$anonfun$receive$1.applyOrElse(Endpoint.scala:287) ~[akka-remote_2.11-2.4.2.jar:na]
        ... 11 common frames omitted
13:15:43.763 [sgActors-akka.actor.default-dispatcher-53] WARN  a.r.Remoting: Association to [akka.tcp://sgActors@app-2016-03-11-130243.as.performance-tests.groupl.es:2552] having UID [-1166429728] is irrecoverably failed. UID is now quarantined and all messages to this
UID will be delivered to dead letters. Remote actorsystem must be restarted to recover from this situation.
13:15:43.764 [sgActors-akka.actor.default-dispatcher-53] WARN  m.s.l.r.QuarantineListener: Actorsystem akka.tcp://sgActors@app-2016-03-11-130243.as.performance-tests.groupl.es:2552 has been quarantined, please restart it
13:15:44.115 [sgActors-akka.actor.default-dispatcher-23] ERROR a.r.Remoting: Association to [akka.tcp://sgActors@app-2016-03-11-130243.as.performance-tests.groupl.es:2552] with UID [-1984549662] irrecoverably failed. Quarantining address.
java.lang.IllegalStateException: Error encountered while processing system message acknowledgement buffer: [0 {0}] ack: ACK[9, {}]
        at akka.remote.ReliableDeliverySupervisor$$anonfun$receive$1.applyOrElse(Endpoint.scala:291) ~[akka-remote_2.11-2.4.2.jar:na]
        at akka.actor.Actor$class.aroundReceive(Actor.scala:480) ~[akka-actor_2.11-2.4.2.jar:na]
        at akka.remote.ReliableDeliverySupervisor.aroundReceive(Endpoint.scala:192) ~[akka-remote_2.11-2.4.2.jar:na]
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526) [akka-actor_2.11-2.4.2.jar:na]
        at akka.actor.ActorCell.invoke(ActorCell.scala:495) [akka-actor_2.11-2.4.2.jar:na]
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257) [akka-actor_2.11-2.4.2.jar:na]
        at akka.dispatch.Mailbox.run(Mailbox.scala:224) [akka-actor_2.11-2.4.2.jar:na]
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234) [akka-actor_2.11-2.4.2.jar:na]
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [scala-library-2.11.7.jar:na]
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [scala-library-2.11.7.jar:na]
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [scala-library-2.11.7.jar:na]
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [scala-library-2.11.7.jar:na]
Caused by: java.lang.IllegalArgumentException: Highest SEQ so far was 0 but cumulative ACK is 9
        at akka.remote.AckedSendBuffer.acknowledge(AckedDelivery.scala:103) ~[akka-remote_2.11-2.4.2.jar:na]
        at akka.remote.ReliableDeliverySupervisor$$anonfun$receive$1.applyOrElse(Endpoint.scala:287) ~[akka-remote_2.11-2.4.2.jar:na]
        ... 11 common frames omitted
13:15:44.115 [sgActors-akka.actor.default-dispatcher-23] WARN  a.r.Remoting: Association to [akka.tcp://sgActors@app-2016-03-11-130243.as.performance-tests.groupl.es:2552] having UID [-1984549662] is irrecoverably failed. UID is now quarantined and all messages to this
UID will be delivered to dead letters. Remote actorsystem must be restarted to recover from this situation.
```

lots of quarantines in effect

**Comment by marekzebrowski** [comment]
[broken-multi-node-join.log.gz](https://github.com/akka/akka/files/169149/broken-multi-node-join.log.gz)
Here is a longer log from multiple nodes trying to join and become quarantined

**Comment by marekzebrowski** [comment]
We did another experiment
1. started cluster with 3 nodes.
2. we spun 2 new nodes and waited until they register
3. we repeated 2) 5 times, so we go 13 nodes in the cluster (3 original and 5 times joined by 2 nodes)
this time everything went well.
It seems that trying to register large(10)  number of new nodes at once  failed in our case

**Comment by patriknw** [comment]
Thanks a lot, we will look into this on monday and tuesday

**Comment by patriknw** [comment]
I was able to reproduce it

**Comment by patriknw** [comment]
This is the reproducer test case:

1. start node1 and actor A
1. start node2 and actor B
1. watch A from B
1. sleep in node1 before sending the Ack for the watch
1. shutdown node2 and start node2' with same host:port
1. node2' will receive the Ack that was targeted for node2, i.e. `Error encountered while processing system message acknowledgement buffer: [-1 {}] ack: ACK[0, {}]`

**Comment by patriknw** [comment]
fixed in master and release-2.3
